{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "40fae134",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a10c48f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex.id</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>mhc.class</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>...</th>\n",
       "      <th>antigen.species</th>\n",
       "      <th>reference.id</th>\n",
       "      <th>method</th>\n",
       "      <th>meta</th>\n",
       "      <th>cdr3fix</th>\n",
       "      <th>vdjdb.score</th>\n",
       "      <th>web.method</th>\n",
       "      <th>web.method.seq</th>\n",
       "      <th>web.cdr3fix.nc</th>\n",
       "      <th>web.cdr3fix.unmp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEAGQGFFSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSFEAGQGFFSNQPQHF\", \"cdr3_old\": \"C...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   complex.id gene                  cdr3       v.segm      j.segm  \\\n",
       "0           1  TRA         CIVRAPGRADMRF  TRAV26-1*01   TRAJ43*01   \n",
       "1           1  TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "2           0  TRB   CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "3           2  TRA        CAVPSGAGSYQLTF    TRAV20*01   TRAJ28*01   \n",
       "4           2  TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "\n",
       "       species     mhc.a mhc.b mhc.class antigen.epitope  ... antigen.species  \\\n",
       "0  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "1  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "2  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "3  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "4  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "\n",
       "    reference.id                                             method  \\\n",
       "0  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "1  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "2  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "3  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "4  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "\n",
       "                                                meta  \\\n",
       "0  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "1  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "2  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "3  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "4  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "\n",
       "                                             cdr3fix vdjdb.score  web.method  \\\n",
       "0  {\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...           2        sort   \n",
       "1  {\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...           2        sort   \n",
       "2  {\"cdr3\": \"CASSFEAGQGFFSNQPQHF\", \"cdr3_old\": \"C...           2        sort   \n",
       "3  {\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...           2        sort   \n",
       "4  {\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...           2        sort   \n",
       "\n",
       "  web.method.seq web.cdr3fix.nc web.cdr3fix.unmp  \n",
       "0         sanger             no               no  \n",
       "1         sanger             no               no  \n",
       "2         sanger             no               no  \n",
       "3         sanger             no               no  \n",
       "4         sanger             no               no  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = 'vdjdb.txt' \n",
    "df = pd.read_csv(file_path, sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b7fc746b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# complex.id: 数据记录的复合体ID，用于唯一标识数据记录。\n",
    "# gene: TCR的基因类型，如TRA（α链）或TRB（β链）。\n",
    "# cdr3: TCR中CDR3区的氨基酸序列，这是决定TCR特异性的关键区域。\n",
    "# v.segm: 变异（V）区基因段的名称，对应于TCRα或β链的V区。\n",
    "# j.segm: 连接（J）区基因段的名称，对应于TCRα或β链的J区。\n",
    "# species: TCR来源的物种，如HomoSapiens（人类）。\n",
    "# mhc.a: 主要组织相容性复合体（MHC）分子的A链类型。\n",
    "# mhc.b: MHC分子的B链类型或其他相关标记（如B2M，β2-微球蛋白）。\n",
    "# mhc.class: MHC分子的类别，如MHCI或MHCII。\n",
    "# antigen.epitope: TCR识别的抗原表位的氨基酸序列。\n",
    "# antigen.gene: 抗原的基因名称。\n",
    "# antigen.species: 抗原来源的物种，如HIV-1。\n",
    "# reference.id: 发表这一数据的文献的参考ID，如PMID:15596521。\n",
    "# method: 数据收集的方法或技术，例如通过tetramer-sort、sanger测序等。\n",
    "# meta: 元数据，包含有关样本的额外信息，如细胞亚群、捐献者MHC信息、样本来源等。\n",
    "# cdr3fix: CDR3区序列修正的信息，包括是否需要修正、修正前后的序列等。\n",
    "# vdjdb.score: 数据记录的质量评分。\n",
    "# web.method: 数据提交或验证时使用的网页方法。\n",
    "# web.method.seq: 与web.method相关的序列信息。\n",
    "# web.cdr3fix.nc: 网络上CDR3修正的非规范（non-canonical）信息。\n",
    "# web.cdr3fix.unmp: 网络上CDR3修正的未映射（unmapped）信息。\n",
    "# complex.id: Complex ID of the data record, used to uniquely identify the data record.\n",
    "# gene: The gene type of the TCR, such as TRA (alpha chain) or TRB (beta chain).\n",
    "# cdr3: Amino acid sequence of the CDR3 region in the TCR, which is the key region that determines the specificity of the TCR.\n",
    "# v.segm: Name of the gene segment of the variant (V) region, corresponding to the V region of the TCR alpha or beta chain.\n",
    "# j.segm: Name of the gene segment of the junction (J) region, corresponding to the J region of the TCR alpha or beta chain.\n",
    "# species: The species from which the TCR originated, e.g. HomoSapiens (human).\n",
    "# mhc.a: A-chain type of the major histocompatibility complex (MHC) molecule.\n",
    "# mhc.b: B-chain type of MHC molecule or other relevant marker (e.g. B2M, β2-microglobulin).\n",
    "# mhc.class: Class of the MHC molecule, e.g. MHCI or MHCII.\n",
    "# antigen.epitope: Amino acid sequence of the antigenic epitope recognised by the TCR.\n",
    "# antigen.gene: The gene name of the antigen.\n",
    "# antigen.species: The species from which the antigen originated, e.g. HIV-1.\n",
    "# reference.id: The reference ID of the literature in which this data was published, e.g. PMID:15596521.\n",
    "# method: The method or technique used for data collection, e.g. by tetramer-sort, sanger sequencing, etc.\n",
    "# meta: Metadata containing additional information about the sample, e.g. cell subpopulations, donor MHC information, sample source, etc.\n",
    "# cdr3fix: information about sequence correction of CDR3 region, including whether correction is needed, sequence before and after correction, etc.\n",
    "# vdjdb.score: Quality score of the data record.\n",
    "# web.method: The web method used for data submission or validation.\n",
    "# web.method.seq: Sequence information associated with the web.method.\n",
    "# web.cdr3fix.nc: Non-canonical information about CDR3 fixes on the web.\n",
    "# web.cdr3fix.unmp: Unmapped information for CDR3 fixes on the web."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bfb23766",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.class</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>antigen.gene</th>\n",
       "      <th>antigen.species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>Nef</td>\n",
       "      <td>HIV-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>Nef</td>\n",
       "      <td>HIV-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEAGQGFFSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>Nef</td>\n",
       "      <td>HIV-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>Nef</td>\n",
       "      <td>HIV-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>Nef</td>\n",
       "      <td>HIV-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene                  cdr3       v.segm      j.segm      species mhc.class  \\\n",
       "0  TRA         CIVRAPGRADMRF  TRAV26-1*01   TRAJ43*01  HomoSapiens      MHCI   \n",
       "1  TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01  HomoSapiens      MHCI   \n",
       "2  TRB   CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01  HomoSapiens      MHCI   \n",
       "3  TRA        CAVPSGAGSYQLTF    TRAV20*01   TRAJ28*01  HomoSapiens      MHCI   \n",
       "4  TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01  HomoSapiens      MHCI   \n",
       "\n",
       "  antigen.epitope antigen.gene antigen.species  \n",
       "0        FLKEKGGL          Nef           HIV-1  \n",
       "1        FLKEKGGL          Nef           HIV-1  \n",
       "2        FLKEKGGL          Nef           HIV-1  \n",
       "3        FLKEKGGL          Nef           HIV-1  \n",
       "4        FLKEKGGL          Nef           HIV-1  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select the relevant column\n",
    "relevant_columns = [\n",
    "    'gene', 'cdr3', 'v.segm', 'j.segm', 'species',\n",
    "    'mhc.class', 'antigen.epitope', 'antigen.gene', 'antigen.species'\n",
    "]\n",
    "df = df[relevant_columns]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a2c357d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.class</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>antigen.gene</th>\n",
       "      <th>antigen.species</th>\n",
       "      <th>cdr3_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>0</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>Nef</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>0</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>Nef</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEAGQGFFSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>0</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>Nef</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>0</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>Nef</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>0</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>Nef</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene                  cdr3       v.segm      j.segm      species  mhc.class  \\\n",
       "0  TRA         CIVRAPGRADMRF  TRAV26-1*01   TRAJ43*01  HomoSapiens          0   \n",
       "1  TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01  HomoSapiens          0   \n",
       "2  TRB   CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01  HomoSapiens          0   \n",
       "3  TRA        CAVPSGAGSYQLTF    TRAV20*01   TRAJ28*01  HomoSapiens          0   \n",
       "4  TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01  HomoSapiens          0   \n",
       "\n",
       "  antigen.epitope antigen.gene antigen.species  cdr3_length  \n",
       "0        FLKEKGGL          Nef           HIV-1           13  \n",
       "1        FLKEKGGL          Nef           HIV-1           20  \n",
       "2        FLKEKGGL          Nef           HIV-1           19  \n",
       "3        FLKEKGGL          Nef           HIV-1           14  \n",
       "4        FLKEKGGL          Nef           HIV-1           19  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Data Cleaning\n",
    "df = df.dropna()\n",
    "\n",
    "# Feature engineering\n",
    "df['cdr3_length'] = df['cdr3'].apply(len)\n",
    "\n",
    "# Data conversion\n",
    "df['mhc.class'] = df['mhc.class'].astype('category').cat.codes\n",
    "\n",
    "# Showing the first few rows of data after preprocessing\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15d37f81",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The main limitation of using the distance or similarity score matrix between TCR sequences to predict specificity \n",
    "# in the method you provided is the inconsistency in the length of the CDR3 region. Since the CDR3 regions of TCR \n",
    "# sequences can have different lengths, this complicates direct comparisons or the creation of uniform feature \n",
    "# representations. Solo thermal coding produces highly sparse data when dealing with this type of problem, which \n",
    "# may be detrimental to model training and efficiency. In addition, simple solo thermal coding may not be able to \n",
    "# capture the complex interactions and biological properties between amino acids that are critical for determining \n",
    "# the specificity of TCRs.\n",
    "\n",
    "# To overcome these limitations, several strategies can be employed. Firstly, the length of the CDR3 region can be \n",
    "# standardised by padding or truncating sequences to ensure that all sequences have the same dimension when encoded. \n",
    "# Second, other biological features of the sequences, such as the physicochemical properties of amino acids, can be \n",
    "# introduced, as well as the use of advanced sequence encoding methods, such as word embedding techniques, to reduce \n",
    "# sparsity and provide a richer representation of the sequences. In addition, the use of deep learning models to \n",
    "# process variable-length sequences and learn complex patterns of sequences, or sequence alignment algorithms to \n",
    "# ensure accuracy of comparisons can be considered. With these approaches, the performance of prediction models can \n",
    "# be improved and the specificity of TCRs can be predicted more accurately.\n",
    "\n",
    "# 在您提供的方法中，利用TCR序列之间的距离或相似度分数矩阵来预测特异性的主要局限性在于CDR3区域长度的不一致性。由于TCR序列的CDR3区域可以有不\n",
    "# 同的长度，这使得直接比较或创建统一的特征表示变得复杂。独热编码在处理这类问题时会产生高度稀疏的数据，这可能不利于模型的训练和效率。此外，简单\n",
    "# 的独热编码可能无法捕捉氨基酸之间的复杂相互作用和生物学特性，这些特性对于确定TCR的特异性至关重要。\n",
    "\n",
    "# 为了克服这些局限性，可以采取多种策略。首先，可以通过对序列进行填充或截断来标准化CDR3区域的长度，以确保所有序列在编码时具有相同的维度。其次\n",
    "# ，可以引入序列的其他生物学特征，如氨基酸的物理化学属性，以及使用高级的序列编码方法，如词嵌入技术，以降低稀疏性并提供更丰富的序列表示。此外\n",
    "# ，还可以考虑使用深度学习模型来处理变长序列并学习序列的复杂模式，或者采用序列对齐算法来确保比较的准确性。通过这些方法，可以提高预测模型的性\n",
    "# 能，更准确地预测TCR的特异性。\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c20d493",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
